Added a new function for a hardcoded filter of namespaces

xiaoyu74 opened the pull request:
### Which issue this PR addresses:

Implemented a hardcoded filter of namespace: Check the [ADO work item](https://msazure.visualstudio.com/AzureRedHatOpenShift/_boards/board/t/Skippy/Stories/?workitem=12682885) for the details

### What this PR does / why we need it:

- We use the filtering in code in ARO-RP to reduce metrics generated and send to Geneva, and inside Geneva Monitoring to reduce the number of customer-generated-alerts
- Since the list of namespaces is not changing often and there is a chance of unwanted namespaces to be included in the list if we autogenerate it from a release, so it makes sense to hardcode the list in go code for now.
- The list of the hardcoded namespaces are mainly based on the items in the [ConfigMap]( https://github.com/openshift/managed-cluster-config/blob/ade1697fd61ff033f8fe24134df6bba3a390a000/deploy/osd-managed-resources/ocp-namespaces.ConfigMap.yaml) and 2 additional namespaces `openshift-azure-operator` and `openshift-managed-upgrade-operator`. 

### Test plan for issue:
Added an unit test for the new function of hardcoded filter in `namespace_test.go`

### Is there any documentation that needs to be updated for this PR?

No, this PR is based on an [ADO work item](https://msazure.visualstudio.com/AzureRedHatOpenShift/_boards/board/t/Skippy/Stories/?workitem=12682885) we have discussed

> > Let's do something more intelligent for ensuring that when we release a new minor version, we're including the newest namespaces that upstream OCP may be introducing.
> > Write an e2e test which fetches all the namespaces within a newly created cluster, and then ensures that each namespace (with the sole exception of default) returns true from this function.
> 
> @bennerv I wanted this work item to be a quick win so we have better filtering. I believe namespaces do not change often, but I agree - it is a good idea to update `upstream-differences.md`. Let's look at this as a first iteration. So in this PR I would like to see:
> 
> * More precise filtering in the simplest way possible
> * Updated vendoring documentation
> 
> Next iteration could be something like this:
> 
> * Extract namespaces from a OCP release (similar to what OSD is doing [here](https://github.com/openshift/managed-cluster-config/blob/1d5397c516289da8f3ef4fc4c8ab306566ba8b8a/scripts/managed-resources/generate-managed-list.py#L44-L75)) and mix them with ARO-specific namespaces.
> * Update the filtering to use generated list for filtering
> 
> The tricky bit here is that we will soon-ish be looking into adding support of multiple OCP versions via Hive and I don't know how it is going to work just yet. So probably it is not worth implementing extraction right now.

@m1kola - I just created another draft PR based on the review feedback, since you created the work item and should know more context, please help to review first to see if it's what you need, then I can I request other teammates to review further

bryanro92 commented:
/azp run ci

bryanro92 commented:
/azp run e2e

bennerv commented on pkg/util/namespace/namespace.go:
```diff
@@ -23,3 +23,69 @@ func IsOpenShiftSystemNamespace(ns string) bool {
+		"openshift-azure-operator":                         true,
+		"openshift-managed-upgrade-operator":               true,
+	}
+	return filtered_namespaces[ns]
```
We should use a `map[string]struct{}` here instead of `map[string]bool` as golang has a construct when accessing values out of a map to determine if they exist.  We use this pattern in a few places. 

https://github.com/Azure/ARO-RP/blob/master/cmd/aro/mirror.go#L25-L28
https://github.com/Azure/ARO-RP/blob/master/cmd/aro/mirror.go#L112-L114

you can then do something like

``` 
_, ok := filtered_namespaces[ns]
return ok
```

xiaoyu74 replied:
Thanks Ben, I have updated the code based on your suggestions in the new commit.

bennerv commented on pkg/util/namespace/namespace.go:
```diff
@@ -23,3 +23,69 @@ func IsOpenShiftSystemNamespace(ns string) bool {
+
+// FilteredOpenShiftNamespace returns true if ns is a namespace in the defined hardcoded map
+func FilteredOpenShiftNamespace(ns string) bool {
+	filtered_namespaces := map[string]bool{
```
typical variable usage is golang is camelcase
```suggestion
	filteredNamespaces := map[string]bool{
```

xiaoyu74 replied:
Thanks Ben. I have updated the variable name using the pattern you suggested.

bennerv reviewed (changes requested):
Few nits.  

Additionally, we will have to update our vendoring documentation and ensure that we update this list of namespaces whenever there is a new minor version.  Every minor version upstream OCP seems to release a new namespace or two.

xiaoyu74 commented:
> Few nits.
> 
> Additionally, we will have to update our vendoring documentation and ensure that we update this list of namespaces whenever there is a new minor version. Every minor version upstream OCP seems to release a new namespace or two.

Sorry, what's the vendoring documentation and how should I update it ?

bennerv commented on pkg/util/namespace/namespace.go:
```diff
@@ -23,3 +23,70 @@ func IsOpenShiftSystemNamespace(ns string) bool {
+
+// FilteredOpenShiftNamespace returns true if ns is a namespace in the defined hardcoded map
+func FilteredOpenShiftNamespace(ns string) bool {
+	filteredNamespaces := map[string]struct{}{
```
can you put these in alphabetical order as well?

xiaoyu74 replied:
Hi Ben, thanks for your review feedback, as I mentioned in another reply, we probably need to get some input from @m1kola to clarify the requirement of the work item.

facchettos replied:
It's not the first time I'm seeing this pattern in the repo, so there's probably a reason behind it , but why not use a ```map[string]bool``` instead of ```map[string]struct{}``` ? It seems like the former is more readable and avoids a 
```
_, ok := filteredNamespaces[ns]
return ok

s-amann replied:
The boolean value in the map doesn't provide any additional value to determining if a key exists. Since Golang doesn't have `contains` on an array or slice the use of an empty struct takes up less memory. 

There is this discussion about it which goes deeper: https://gist.github.com/davecheney/3be245c92b61e5045f75 and if you want to read some funny comments on how go doesn't have "array lookups" -> https://stackoverflow.com/questions/10485743/contains-method-for-a-slice

TL;DR, it doesn't really make a huge difference. If we decide `map[string]bool` is easier to read we should have a refactor story to remain consistent in this codebase.

facchettos replied:
But you could still do ```isfiltered, ok := filteredNamespaces[ns]``` even with booleans if you really want to see if a value has been set. Here we are effectively using this as a boolean map, because the only set values are the excluded namespaces.  

Also if we ever want to exclude a namespace from filtering for some reason, we could explicitly set it to false instead of just not putting it in the list.

xiaoyu74 replied:
@facchettos - Thanks for your time to review. Yeah, I used the `map[string]bool` in my first commit but got review feedback suggested that I should follow the previous pattern of the codebase to keep it consistent. Since it doesn't make big difference so I agreed to use `map[string]struct{}` for now.

bennerv replied:
@facchettos I agree maybe we should create a work item to convert everything.  For now I think it's fine to leave the changes as @xiaoyu74 has them, but we can convert all `map[string]struct{}` to `map[string]bool` in a follow-up item.

bennerv reviewed (changes requested):
Let's do something more intelligent for ensuring that when we release a new minor version, we're including the newest namespaces that upstream OCP may be introducing. 

Write an e2e test which fetches all the namespaces within a newly created cluster, and then ensures that each namespace (with the sole exception of `default`) returns true from this function.  

That means that when we introduce new minor versions, another namespace, etc. we will cover our butts on this.

bennerv commented:
> Sorry, what's the vendoring documentation and how should I update it ?

https://github.com/Azure/ARO-RP/blob/master/docs/upstream-differences.md

bennerv commented:
One more comment.

We're introducing a new function, but not actually using it anywhere.  Does this story include the work to utilize this new function as well? If so, please make the appropriate updates here as well.

s-amann commented on pkg/util/namespace/namespace.go:
```diff
@@ -23,3 +23,70 @@ func IsOpenShiftSystemNamespace(ns string) bool {
+		"openshift-user-workload-monitoring":               {},
+		"openshift-vsphere-infra":                          {},
+		"openshift-azure-operator":                         {},
+		"openshift-managed-upgrade-operator":               {},
```
I think we need to add `openshift-azure-logging` to this list, if this is to be used in the same places as `IsOpenshiftSystemNamespace` we would lose our monitoring for our logging/mdsd pods.

m1kola replied:
+1. I would also visually separate ARO-specific namespaces from OCP namespaces. So it is easier to update this map.

bryanro92 commented on pkg/util/namespace/namespace.go:
```diff
@@ -23,3 +23,70 @@ func IsOpenShiftSystemNamespace(ns string) bool {
 }
+
+// FilteredOpenShiftNamespace returns true if ns is a namespace in the defined hardcoded map
+func FilteredOpenShiftNamespace(ns string) bool {
```
As I understand it, this function was created because the prefix `openshift-` here https://github.com/Azure/ARO-RP/pull/1994/files#diff-e6fbfa5cca3e92386cc833c67dea826373fc746278df21f8e778d13d9ed6b103R24 can capture customer `openshift-*` namespaces such as `openshift-gitops`.

Does it make sense to fix the previous function, or add an entirely new one? As I saw in a previous review, this new function is not actually being called or performing any work at this point. What was the plan with this moving forward?

If we just patched the functionality of `IsOpenShiftSystemNamespace` to include each namespace we care about granularly this new function would not be required.

xiaoyu74 replied:
Hi Ben, thanks for your detailed review feedback. 

Based on the discussion with @m1kola in the [ADO - 12682885](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12682885). - He suggested we would go with hard-coding for now. Extracting namespaces from an OCP release could be a next iteration. 

That's why I just added a new function "FilteredOpenShiftNamespace" and still keep the previous functions at this stage.

m1kola replied:
Sorry if it wasn't clear from the work item, but the idea is to replace existing filtering with new, more accurate. Also take a look at the acceptance creteria of the work item: this PR doesn't meet it as we are not using the new filtering at the moment.

darthhexx replied:
Originally there were 2 distinct functions, so the name `FilteredOpenShiftNamespace` sort of made sense, but now that it has replaced `IsOpenShift` and `IsOpenShiftSystemNamespace` entirely I would suggest that the function name should be changed across the board to `IsOpenShiftNamespace`.

xiaoyu74 commented:
> One more comment.
> 
> We're introducing a new function, but not actually using it anywhere. Does this story include the work to utilize this new function as well? If so, please make the appropriate updates here as well.

I think this [work-item-12682885](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12682885) is just to implement a new function of a hardcoded filter of namespaces, @m1kola should have another work item which depends on it, will double check with him.

m1kola reviewed (changes requested):
> Let's do something more intelligent for ensuring that when we release a new minor version, we're including the newest namespaces that upstream OCP may be introducing.
>
> Write an e2e test which fetches all the namespaces within a newly created cluster, and then ensures that each namespace (with the sole exception of default) returns true from this function.

@bennerv I wanted this work item to be a quick win so we have better filtering. I believe namespaces do not change often, but I agree - it is a good idea to update `upstream-differences.md`. Let's look at this as a first iteration. So in this PR I would like to see:

* More precise filtering in the simplest way possible
* Updated vendoring documentation

Next iteration could be something like this:

* Extract namespaces from a OCP release (similar to what OSD is doing [here](https://github.com/openshift/managed-cluster-config/blob/1d5397c516289da8f3ef4fc4c8ab306566ba8b8a/scripts/managed-resources/generate-managed-list.py#L44-L75)) and mix them with ARO-specific namespaces.
* Update the filtering to use generated list for filtering

The tricky bit here is that we will soon-ish be looking into adding support of multiple OCP versions via Hive and I don't know how it is going to work just yet. So probably it is not worth implementing extraction right now.

xiaoyu74 commented:
> Could you please move changes into #1994 and close this PR? We had quite a lot of discussions around how this should be implemented in the original PR and preserving this history might be useful for the future iterations around filtering.

Hi @m1kola , do you want to me to move the changes from the draft PR #1999 to this #1994 and close this #1994 which means we don't need to implement this hardcoded filter of namespaces for now ?

m1kola reviewed (changes requested):
I see that CI is failing. I think this is due to the fact that places where we were calling `IsOpenShift` where not updated.

I did not review the PR on this iteration: will come back once CI is green.

xiaoyu74 commented:
Submitted a new commit based on @m1kola clarifications, just a brief summary about the new commit:
- Implemented a more precise hardcoded filtering function with the specified namespaces we care about
- Removed the previous filtering functions in the namespace.go and updated the unit test.
- Updated the vendoring documentation
- Applied the new filtering function to the monitor

xiaoyu74 commented:
@m1kola - As you mentioned we will need to replace the previous filtering functions with the new I created which means I need to do some extra work to update the unit tests and other monitoring code rely on the new function to make the CI pipeline works. (I used to just appended the new function and keep the previous functions so we don't have to update anything as I remember you want to use the new function in the next iteration). 

So, it's still what we want to replace the previous filtering right ? So I will update the code relying on the new function accordingly.

xiaoyu74 commented:
The CI pipeline has passed. Need to do some debugging on the E2E test failed

xiaoyu74 commented:
@m1kola - I have replaced the filtering with the new hard-coded one and fixed the CI and E2E test issues and it's all passed now. Please feel free to review. 

One thing I need to your help is to check if the current hard-coded list of the namespaces are all what we exactly care about or we need to append some more. Thanks.

m1kola commented on docs/upstream-differences.md:
````diff
@@ -88,3 +88,70 @@ Once installer fork is ready:
+
+The namespaces are put in alphabetical order below: 
+```
+openshift
````
I would remvoe the list of namespaces the the docs: we don't want to update two places (code and docs) manually.

But we should add a step under "Update ARO-RP" saying that the list in `pkg/util/namespace/namespace.go` needs to be updated as part of introduction of a new minor OCP version as suggedted by Ben [here](https://github.com/Azure/ARO-RP/pull/1994#pullrequestreview-900381460).

m1kola commented on pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go:
```diff
@@ -225,51 +225,55 @@ func TestValidateAdminKubernetesObjectsNonCustomer(t *testing.T) {
 		{
 			test:      "allowed groupKind on read",
 			groupKind: "ClusterRole.rbac.authorization.k8s.io",
+			namespace: "openshift",
```
I think we introduced a bug somewhere and made tests pass by adding a namespace here and in test cases below.

`ClusterRole` is a non-namespaced kind. Geneva action should not require namespace in this case.

xiaoyu74 replied:
Since the function we are using `validateAdminKubernetesObjectsNonCustomer()` specified the `namespace` as parameter, so we have to pass a namespace to it (It will still pass " " as the namespace if we don't do so but " " is not included in our hard-coded list and will return errors)

So, we can fix it by adding the " " in our hard-coded list if it's what we want (The previous filtering `IsOpenshift()` did add " ") or update the `validateAdminKubernetesObjectsNonCustomer()` but it should be a separated issue from this PR.

m1kola commented on pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go:
```diff
@@ -278,27 +282,27 @@ func TestValidateAdminKubernetesObjectsNonCustomer(t *testing.T) {
 			namespace: "openshift-/",
 			name:      "Valid-NAME-01",
-			wantErr:   "400: InvalidParameter: : The provided namespace 'openshift-/' is invalid.",
+			wantErr:   "403: Forbidden: : Access to the provided namespace 'openshift-/' is forbidden.",
```
Why did it change from 400 to 403?

xiaoyu74 replied:
Since we hard-coded the namespaces without any invalid name matching the regex in the `validate.go`, so the function `validateAdminKubernetesObjects()` will never be called and return  `400: InvalidParameter: : The provided namespace 'openshift-/' is invalid.` but only return `403: Forbidden: : Access to the provided namespace 'openshift-/' is forbidden.` from the function `validateAdminKubernetesObjectsNonCustomer()`

m1kola commented on test/e2e/adminapi_kubernetesobjects.go:
```diff
@@ -26,7 +26,7 @@ var _ = Describe("[Admin API] Kubernetes objects action", func() {
 	const objName = "e2e-test-object"
 
 	When("in a standard openshift namespace", func() {
-		const namespace = "default"
```
It looks like we are dropping support of the default manespace. I know that customers often deploy their workloads there (probably not intentionally in many cases).

Is there anything important for us to monitor in there? I don't think so.

@gvanderpotte and other RLs. Any thoughts?

bennerv replied:
We don't deploy pods in default so IMO we should drop it.  It's a standard namespace in a sense that it exists in an openshift cluster, but besides that it's of no use to us.

SudoBrendan replied:
I'm unsure of what exactly we mean here when we say "dropping support"... but I know that generally, there are two very important objects that are expected to exist in `default` from a Kubernetes perspective, and we see that in OpenShift/ARO clusters too:

- `service/kubernetes` - a ClusterIP Service used by Pods to hit the K8s API
- `service/openshift` - apparently just an ExternalName record to `service/kubernetes`...

`service/kubernetes` is actually so important, that if you delete it the API automatically recreate it because it's how all controllers interact with the cluster API.

xiaoyu74 replied:
@SudoBrendan - Thanks for your comment. So the problem is what exactly the namespaces we really care about. If we only care about the namespaces that the customer deploy their workloads to, then we probably don't have to keep the `default` and the other 3 initial namespaces as below: 
- `kube-system`
- `kube-public`
- `kube-node-lease`

However, if we want to get monitoring from the kubernetes system level for auditing purpose (Checking the API traffic that the pods sending requests to the Kubernetes master) Then, it makes sense to keep the `default` and the `kube-` namespaces. Also, the previous function `IsOpenShift` and `IsOpenShiftSystemNamespace` also handled the `default` and `kube-` namespaces. Our new filtering function should support the same as before ? How do you think @bennerv @m1kola ?

m1kola replied:
I think there is some confusion in this thread.

We collect the following things:

* Logs (audit logs and logs from OCP workloads). It is a separate config within the cluster.
    Here is where we define namespaces from which we gather logs:
    https://github.com/Azure/ARO-RP/blob/6c9de0ae5c672e6c4323d0ca5f7e55dbeeb5d124/pkg/operator/controllers/genevalogging/const.go#L73-L76
    [This file](https://github.com/Azure/ARO-RP/blob/6c9de0ae5c672e6c4323d0ca5f7e55dbeeb5d124/pkg/operator/controllers/genevalogging/const.go) contains other log related configs (including audit logs).
    **Note:** this PR does not affect logs.
* Metrics. These are gathered by monitor component (`pkg/monitor`). This is what we are changing in this PR.

---

I do not think we need `default`, `kube-*` namespaces here. We do not deploy any OCP things we gather metrics on into these namespaces as far as I know. But customers can deploy pods, jobs, etc into these namespaces and it will introduce noise into our metrics.

m1kola commented on pkg/util/namespace/namespace_test.go:
```diff
@@ -7,62 +7,30 @@ import (
 		{
-			want: true,
-		},
-		{
-			namespace: "openshift-ns",
-			want:      true,
-		},
```
Instead of removing test cases I would update them to reflect new filtering. So this one becomes `want: false`.

m1kola commented on pkg/util/namespace/namespace_test.go:
```diff
@@ -7,62 +7,30 @@ import (
 		namespace string
 		want      bool
 	}{
 		{
-			want: true,
-		},
```
I think we still need this test case. Removing it allowed regression with non-namespaced resources to happen.

m1kola reviewed (changes requested):
I think there is an issue with non-namespaced resources such as `ClusterRole` which needs to be fixed.

Also I'm not sure about dropping "default" namespace. Would love to hear more opinions on that.

xiaoyu74 commented:
@m1kola @gvanderpotte - Any updates about this PR ?  Looking forward to getting your feedback. Thanks

darthhexx commented on pkg/util/namespace/namespace.go:
```diff
@@ -3,23 +3,70 @@ package namespace
+// FilteredOpenShiftNamespace returns true if ns is a namespace in the defined hardcoded map
+func FilteredOpenShiftNamespace(ns string) bool {
+	filteredNamespaces := map[string]struct{}{
+		"":                                                 {},
```
As mentioned in other comments in this PR, please would you separate the ARO and OCP namespaces so our future selves can more easily curate the lists. Also, please add a comment about the empty "" namespace so it's clear why it's necessary.

Separately, when compared to the previous functions definition, I would say this list is also missing the `kube-system` namespace.

xiaoyu74 replied:
@darthhexx - Thanks for your review feedback.  I hard-coded the namespaces mainly based on the [configMap](https://github.com/openshift/managed-cluster-config/blob/ade1697fd61ff033f8fe24134df6bba3a390a000/deploy/osd-managed-resources/ocp-namespaces.ConfigMap.yaml) and it doesn't contain `kube-system` namespace. 

Do we really need to add kube-system related namespace? If so, what particular `kube-system` namespace we should hard-code? The previous functions just check namespaces by patterns `strings.HasPrefix(ns, "kube-")` but I wasn't aware in the ADO task that the kube-system needs to be included ?

darthhexx replied:
I haven't checked all of the places that the function is called, but the original functions would have matched on `kube-system`, which we do interact with. I was merely highlighting the fact that the new function is not only being stricter about openshift namespaces but also dropping kube-* namespace support. Is this intended?

xiaoyu74 replied:
@m1kola - As David mentioned above, I also wanted to get clarified that if the `kube-*` namespaces should be hardcoded or we only need to filter `openshift-` namespaces in the new hard-coded function as intended?

darthhexx reviewed (changes requested):
A few changes requested.

ellis-johnson commented on pkg/util/namespace/namespace.go:
```diff
@@ -3,23 +3,73 @@ package namespace
+		"openshift-ovirt-infra":                            {},
+		"openshift-sdn":                                    {},
+		"openshift-service-ca":                             {},
+		"openshift-service-ca-operator":                    {},
```
Will this list of namespaces potentially change over time? I'm just cautious of using this hard-coded list and having to manage it over time

xiaoyu74 replied:
@ellis-johnson - Thanks for reviewing. Yes, as we discussed in the ADO task and this PR conversation, we would like to have a more precise filtering in the simplest way possible and the list of the hard-coded namespaces in `pkg/util/namespace/namespace.go` needs to be updated regularly as every minor version of upstream OCP introduces a new namespace or two. 

In the next iteration, we probably will do a more intelligent way to extract namespaces from an OCP release (similar to what OSD is doing [here](https://github.com/openshift/managed-cluster-config/blob/1d5397c516289da8f3ef4fc4c8ab306566ba8b8a/scripts/managed-resources/generate-managed-list.py#L44-L75)) and mix them with ARO-specific namespaces.

ellis-johnson reviewed (commented):
One comment, but otherwise lgtm!

facchettos commented on pkg/frontend/validate.go:
```diff
@@ -77,7 +77,7 @@ func (f *frontend) validateOpenShiftUniqueKey(ctx context.Context, doc *api.Open
 
 func validateAdminKubernetesObjectsNonCustomer(method, groupKind, namespace, name string) error {
-	if !utilnamespace.IsOpenShift(namespace) {
+	if !utilnamespace.IsOpenShiftNamespace(namespace) {
```
It'd be great to have some consistency on how we import that, why do we import it differently here than from the other files ? Also, maybe I'd call it `namespaceutil` if we are going to import it with that kind of name (I remember some packages in the std library that were named this way.  (e.g. `ioutil` even though it has been deprecated). 
Other solution is to give a less generic name to the package so we don't have conflicts on package names and don't have to rename it. It's up to you 

Apart from that little nit, LGTM

xiaoyu74 replied:
Hi @facchettos - Thanks for your feedback. I think the reason we imported namespace with a different name (`utilnamespace`) is to distinguish with the `namespace` parameter in the function `func validateAdminKubernetesObjectsNonCustomer(method, groupKind, namespace, name string) error`. I am happy to update to use `namespaceutil` if it's a standard naming convention.

facchettos reviewed (changes requested):
Just a little nit. Otherwise LGTM

SudoBrendan commented on pkg/util/namespace/namespace_test.go:
```diff
@@ -7,73 +7,54 @@ import (
 		{
 			namespace: "default",
-			want:      true,
+			want:      false,
```
I left another comment below, but to reiterate:

Namespace `default` houses the cluster-critical `service/kubernetes` and OpenShift-critical `service/openshift` - has this risk been assessed? It might be good to at least keep Events and audit logs for these Services.

m1kola replied:
Just for history. I think this concern was addressed [here](https://github.com/Azure/ARO-RP/pull/1994#discussion_r865019512).

xiaoyu74 commented:
> > Based on that we would stop collecting metrics (and not alert in some cases) on these namespaces:
> 
> Any reason why we're dropping `openshift-authentication`? It comes out of the box and I think we should be collecting on it.
> 
> The PR description says the below. Any reason we want to remove it?
> 
> > openshift-authentication is still included in the hardcoded filter for now.

Sorry for the confusions in the PR description, I created the hardcoded filtering list based on this [ocp-namespace.configmap](https://github.com/openshift/managed-cluster-config/blob/ade1697fd61ff033f8fe24134df6bba3a390a000/deploy/osd-managed-resources/ocp-namespaces.ConfigMap.yaml) at first but noticed that the ADO task [12682885](https://msazure.visualstudio.com/AzureRedHatOpenShift/_boards/board/t/Firefly/Stories/?workitem=12682885) mentioned that the `openshift-authentication` is unwanted.

I have removed the `openshift-authentication` namespace from the filtering list in the previous PR commit. Will update the PR description accordingly.

xiaoyu74 commented:
> Took the liberty collect data on the impact this PR would have on the metrics (well, a subset) we currently collect to help the decision about whether to merge this PR or not. Based on that we would stop collecting metrics (and not alert in some cases) on these namespaces:
> 
> * Default
> * kube-system
> * openshift-authentication
> * openshift-azure-routefix
> * openshift-cloud-network-config-controller
> * openshift-cnv
> * openshift-compliance
> * openshift-dsportal-test
> * openshift-fresh-adminportal-geely
> * openshift-fresh-adminportal-psotest
> * openshift-fresh-adminportal-test1
> * openshift-fresh-dsportal
> * openshift-gitops
> * openshift-local-storage
> * openshift-operators-redhat
> * openshift-pipelines-app-delivery
> * openshift-storage
> * openshift-user-workload-monitoring
> * openshift-workspaces

Hi @UlrichSchlueter - Thanks for your review feedback. I can confirm that all the namespaces you listed above are excluded from the hardcoded filtering I created.  

Just one thing need to double check is that if we should remove `openshift-operator-lifecycle-manager` as we got alert from this namespace which probably is unwanted based on the discussion with our NASA/APAC mates in today's oncall shift.

UlrichSchlueter commented:
Did some more digging into the Geneva side of things. We are touching two things:
1) Metrics and monitoring (that's the whole idea of this PR)
2) The two Geneva action that allow us to path or delete Kubernetes Objects in the filtered list of namespaces (which was probably not intended originally)

About 1) I checked the definitions to see what IcM we would drop.:
- we have three monitors that are metric based (on job.conditions, pod.conditions and on prometheus.alerts metrics).
- The namespaces mentioned in the pod.conditions monitor are all covered by the namespace list.
- For the prometheus alerts we filter by namespace, even though two of the three cases we cover on the Geneva side don't seem to have namespace set at all, which might be the reason why we need the empty value in the namespace list. We don't send the namespace data over to Geneva so no historic data here (Might be a good idea to remote the namespace filter on the RP side in a next step)
- For the job,conditions we have two filters , one on the RP side and one on the Geneva. Several namespace we permit on the RP side are filtered out in the monitor.  ( the geneva side exludes  openshift-logging, openshift-gitops ,openshift-authentication,    openshift-operators, openshift-marketplace ,openshift-pipelines ,  openshift-cluster-version, openshift-image-registry ,   openshift-operators-redhat,     openshift-compliance, and     openshift-multus) Anyways If the new list would have been place since the beginning of generating ICMs for job.conditions we would have received one less IcM (for the openshift-acr namespace) 

All in all on the monitoring side we should be fine with the list  and not drop any important IcMs.

About 2) we introduce more restrictions than we have now which was not originally intended.  Unfortunately I couldn't find a way to check how much impact we would have seen on historic cases. Might make to restore the original behavior for the frontend code.


With all that said my suggestion would be:
- Add openshift-operator-lifecycle-manager to the exclusion list of the jobs.conditions monitor (see V2 monitors)
- Fix the remaining comments ( the 'only add to list, never delete' one )
- Otherwise leave the list as it is (leaving the openshift-authentication out, as per original ticket and purpose)
- Revert the change for the frontend code (reintroducing one of the original functions)
- Merge
- Open a ticket to remove the namespace filter on the prometheus.alerts metric (maybe replace with an alertname filter)
- Once in production cleanup the jobs.condition filter on the Geneva side
- Continue with pods.restartcounter monitor that is waiting for this filter (and which is why I'm looking into this PR in the first place)

@xiaoyu74 @m1kola @bennerv what do you think ?

m1kola reviewed (changes requested):
TLDR: I think we are nearly there with this PR.

IMO the next steps are (based on suggestion from @UlrichSchlueter):

* Address the remaining PR feedback.
    * [The 'only add to list, never delete' comment](https://github.com/Azure/ARO-RP/pull/1994#discussion_r860851774)
    * And exclude `openshift-operator-lifecycle-manager`  from the list (I'm missing background on this decision - assuming this is a correct decision)
    * Comments from my review
* Create a work item to update Geneva monitors to remove temporary exclusions (e.g. on the `job.conditions`, see more details below) once this PR is rolled out.
* Merge
* As a separate work items/PRs (needs to be created) look into potential improvements (e.g. refactoring `""` value and in enforcing convention for import name)

---

@UlrichSchlueter your description is mostly correct.

We are touching two areas:

1. Metrics and monitoring - that was the initial intent of the work item: reduce the noise in metrics and in incidents.
2. The two Geneva action that allow SREs to issue `POST` and `DELETE` Kubernetes Objects. We did not intend to modify this, but it is ok (see more below on this topic).

> we have three monitors that are metric based (on job.conditions, pod.conditions and on prometheus.alerts metrics).

We have three monitors on Geneva side, but we emit metrics which rely on namespace filtering the following places:

* `emitDaemonsetStatuses`
* `emitDeploymentStatuses`
* `emitJobConditions`
* `emitPodConditions` (3 usages down the call)
* `emitReplicasetStatuses`
* `emitStatefulsetStatuses`
* `emitPrometheusAlerts`

> For the prometheus alerts we filter by namespace, even though two of the three cases we cover on the Geneva side don't seem to have namespace set at all, which might be the reason why we need the empty value in the namespace list.

1. We need empty string `""` in `IsOpenShiftNamespace` to allow working with non-namespaced objects via Geneva actions. This should be made more clear in the code comment.
2. We also might need an empty string `""` in `IsOpenShiftNamespace` to be able to forward prometheus alerts which do not have namespace label. But I'm not sure about this one.

I left a separate comment on this topic as part of this review.

> For the job,conditions we have two filters , one on the RP side and one on the Geneva. Several namespace we permit on the RP side are filtered out in the monitor. ( the geneva side exludes openshift-logging, openshift-gitops ,openshift-authentication, openshift-operators, openshift-marketplace ,openshift-pipelines , openshift-cluster-version, openshift-image-registry , openshift-operators-redhat, openshift-compliance, and openshift-multus) Anyways If the new list would have been place since the beginning of generating ICMs for job.conditions we would have received one less IcM (for the openshift-acr namespace)

Filtering on the Geneva side was a quick solution. We should revisit and ideally drop these rules from the Geneva side after rolling out this PR into prod. I suggested above in this comment to create a work item for this.

> All in all on the monitoring side we should be fine with the list and not drop any important IcMs.

+1

> About 2) we introduce more restrictions than we have now which was not originally intended. Unfortunately I couldn't find a way to check how much impact we would have seen on historic cases. Might make to restore the original behavior for the frontend code.

On geneva actions side we are now more specific: instead of saying that SREs can issue `POST`/`DELETE` requests against objects in any `openshift-*` namespace we say: SREs can issue `POST`/`DELETE` requests against objects in namespaces from the list. This wasn't on original intent of this work item, but given that:

1. It keeps implementation simple (no need to maintain two sets of namespaces/rules: one for monitor and one for geneva actions)
2. Being more strict in geneva actions is probably a good idea. We never do `POST`/`DELETE` via Geneva actions in customer namespaces (even if they have `openshift-` prefix) anyway. And we should not do that.

I think we should keep using single `IsOpenShiftNamespace` function for both monitor and for geneva action validation.

xiaoyu74 commented on pkg/util/namespace/namespace.go:
```diff
@@ -3,23 +3,73 @@ package namespace
-)
+// IsOpenShiftNamespace returns true if ns is a namespace in the defined hardcoded map
+func IsOpenShiftNamespace(ns string) bool {
+	GetNamespace := map[string]struct{}{
```
> 2. `IsOpenShiftNamespace` currently creates a map on stack every time we call this function and it sounds like a waste for a hardcoded map of namespaces. Probably worth using `switch-case` here instead.

Thanks for your  suggestions. I think Golang map is just a hash table and it should be more efficient to find the particular string from time complexity's perspective. Although it can sacrifice some performance from space complexity's perspective, it's a really small hardcoded map and I think the cost can be ignored to be honest.

xiaoyu74 commented:
> TLDR: I think we are nearly there with this PR.
> 
> IMO the next steps are (based on suggestion from @UlrichSchlueter):
> 
> * Address the remaining PR feedback.
>   
>   * [The 'only add to list, never delete' comment](https://github.com/Azure/ARO-RP/pull/1994#discussion_r860851774)
>   * And exclude `openshift-operator-lifecycle-manager`  from the list (I'm missing background on this decision - assuming this is a correct decision)
>   * Comments from my review
> * Create a work item to update Geneva monitors to remove temporary exclusions (e.g. on the `job.conditions`, see more details below) once this PR is rolled out.
> * Merge
> * As a separate work items/PRs (needs to be created) look into potential improvements (e.g. refactoring `""` value and in enforcing convention for import name)
> 
> @UlrichSchlueter your description is mostly correct.
> 
> We are touching two areas:
> 
> 1. Metrics and monitoring - that was the initial intent of the work item: reduce the noise in metrics and in incidents.
> 2. The two Geneva action that allow SREs to issue `POST` and `DELETE` Kubernetes Objects. We did not intend to modify this, but it is ok (see more below on this topic).
> 
> > we have three monitors that are metric based (on job.conditions, pod.conditions and on prometheus.alerts metrics).
> 
> We have three monitors on Geneva side, but we emit metrics which rely on namespace filtering the following places:
> 
> * `emitDaemonsetStatuses`
> * `emitDeploymentStatuses`
> * `emitJobConditions`
> * `emitPodConditions` (3 usages down the call)
> * `emitReplicasetStatuses`
> * `emitStatefulsetStatuses`
> * `emitPrometheusAlerts`
> 
> > For the prometheus alerts we filter by namespace, even though two of the three cases we cover on the Geneva side don't seem to have namespace set at all, which might be the reason why we need the empty value in the namespace list.
> 
> 1. We need empty string `""` in `IsOpenShiftNamespace` to allow working with non-namespaced objects via Geneva actions. This should be made more clear in the code comment.
> 2. We also might need an empty string `""` in `IsOpenShiftNamespace` to be able to forward prometheus alerts which do not have namespace label. But I'm not sure about this one.
> 
> I left a separate comment on this topic as part of this review.
> 
> > For the job,conditions we have two filters , one on the RP side and one on the Geneva. Several namespace we permit on the RP side are filtered out in the monitor. ( the geneva side exludes openshift-logging, openshift-gitops ,openshift-authentication, openshift-operators, openshift-marketplace ,openshift-pipelines , openshift-cluster-version, openshift-image-registry , openshift-operators-redhat, openshift-compliance, and openshift-multus) Anyways If the new list would have been place since the beginning of generating ICMs for job.conditions we would have received one less IcM (for the openshift-acr namespace)
> 
> Filtering on the Geneva side was a quick solution. We should revisit and ideally drop these rules from the Geneva side after rolling out this PR into prod. I suggested above in this comment to create a work item for this.
> 
> > All in all on the monitoring side we should be fine with the list and not drop any important IcMs.
> 
> +1
> 
> > About 2) we introduce more restrictions than we have now which was not originally intended. Unfortunately I couldn't find a way to check how much impact we would have seen on historic cases. Might make to restore the original behavior for the frontend code.
> 
> On geneva actions side we are now more specific: instead of saying that SREs can issue `POST`/`DELETE` requests against objects in any `openshift-*` namespace we say: SREs can issue `POST`/`DELETE` requests against objects in namespaces from the list. This wasn't on original intent of this work item, but given that:
> 
> 1. It keeps implementation simple (no need to maintain two sets of namespaces/rules: one for monitor and one for geneva actions)
> 2. Being more strict in geneva actions is probably a good idea. We never do `POST`/`DELETE` via Geneva actions in customer namespaces (even if they have `openshift-` prefix) anyway. And we should not do that.
> 
> I think we should keep using single `IsOpenShiftNamespace` function for both monitor and for geneva action validation.

xiaoyu74 commented:
> TLDR: I think we are nearly there with this PR.
> 
> IMO the next steps are (based on suggestion from @UlrichSchlueter):
> 
> * Address the remaining PR feedback.
>   
>   * [The 'only add to list, never delete' comment](https://github.com/Azure/ARO-RP/pull/1994#discussion_r860851774)
>   * And exclude `openshift-operator-lifecycle-manager`  from the list (I'm missing background on this decision - assuming this is a correct decision)
>   * Comments from my review

I have updated the code based on the PR feedback above in the latest commit.

> * Create a work item to update Geneva monitors to remove temporary exclusions (e.g. on the `job.conditions`, see more details below) once this PR is rolled out.
> * Merge
> * As a separate work items/PRs (needs to be created) look into potential improvements (e.g. refactoring `""` value and in enforcing convention for import name)

I will create work items for the points you mentioned above once the PR is rolled out.

m1kola commented on pkg/util/namespace/namespace_test.go:
```diff
@@ -7,73 +7,58 @@ import (
 		want      bool
 	}{
 		{
-			want: true,
```
I think we still need a test case for an empty namespace `""`.

xiaoyu74 replied:
> It looks like we accidentally dropped a test case for empty string `""` instead of a namespace. We need it untill we implement [this suggestion](https://github.com/Azure/ARO-RP/pull/1994#discussion_r864939378).
> 
> Other than that - looks good to me.

Added a test case for the empty namespace `""`
> 
> @xiaoyu74 could you please create a work item to update geneva monitors before we merge this PR? I want to make sure that we don't forget that.

I have created an ADO work item [14351421](https://msazure.visualstudio.com/AzureRedHatOpenShift/_boards/board/t/Firefly/Stories/?workitem=14351421) based on the discussions.

m1kola reviewed (changes requested):
It looks like we accidentally dropped a test case for empty string `""` instead of a namespace. We need it untill we implement [this suggestion](https://github.com/Azure/ARO-RP/pull/1994#discussion_r864939378).

Other than that - looks good to me.

@xiaoyu74 could you please create a work item to update geneva monitors before we merge this PR? I want to make sure that we don't forget that.

m1kola reviewed (approved):
Looks good to me.

@xiaoyu74 we had a lot of reviewers on this PR so I'll keep it open for another working day or so to give others a chance to take another look. Let's merge on Monday EMEA time unless we receive any change requests.
